src/nn/tensor_collection/visitor.rs

```diff
@@ -16,16 +15,18 @@ pub struct RecursiveWalker<'a, M, F> {
 }
 
 /// Something that can visit [Tensor]s. Used in conjunction with [RecursiveWalker].
-pub trait TensorVisitor<E: Dtype, D: DeviceStorage> {
+pub trait TensorVisitor<E: Dtype, D: Device<E>, E2: Dtype, D2: Device<E2>> {
```

Comment:
I was trying yesterday to figure out how we could move this E2 & D2 out of the trait and capture it in some `TensorVisitor::Output` associated type, but failed to.

Any ideas how we could do that? It would greatly simplify some of the trait definitions/generics if we only needed 1 set. We'd basically only need to handle a second dtype/device in the ToDevice/ToDtype impls of TensorVisitor if we could figure that out

Reply:
I've actually decided to make E2 and D2 associated types of both ModuleVisitor and TensorVisitor to make referring to them when implementing iter_tensors easier.

Reply:
Great work!